### ISSUE TITLE
Support direct and indirect left-recursion (Medeiros et al.)

### ISSUE THREAD

**Issue by maxfischer2781** [pyparsing/pyparsing#287]
PyParsing currently does not support "left-recursion", i.e. recursion without advancing the parsing position. Since left-recursion is the natural way to express certain common grammatical features, such as left-associativity, many grammars require cumbersome workarounds.  
As PyParsing is practically a PEG parser, some approaches from recent literature on left-recursive PEG should be compatible. Notably, since PEG itself is not left-recursive, the exact behaviour of left-recursion depends on the specific approach.

As suggested in ["Left Recursion in Parsing Expression Grammars" by Medeiros, Mascarenhas, Ierusalimschy](https://arxiv.org/abs/1207.0443), left-recursion can be modelled as bounded recursion. This is roughly equivalent to duplicating *variable* clauses into a ladder until the longest match is achieved:
```
# left-recursive PEG definition
E <- E '+' num / num
# bounded recursion PEG definition
E0 <- fail
E1 <- E0 '+' num / num
E2 <- E1 '+' num / num
E3 <- E2 '+' num / num
....
```
The variable clause seems to match PyParsing's ``Forward``. Changing ``Forward.parseImpl`` as described in the paper should allow direct and indirect left-recursion.

---------------

The approach by Medeiros et al. is the basis for [PEP 617: Left recursion](https://www.python.org/dev/peps/pep-0617/#left-recursion). The PEP uses an optimised but complicated solution that relies on re-using a Packrat cache. In order to work with and without ``enablePackrat``, to provide a simple base implementation, and speed up development, a direct implementation seems prudent.

**Comment by maxfischer2781** [comment]
I have just started working on this, but am still far away from having a working implementation and a PR. Feel free to assign this ticket to me.  
I will drop notes here if there are any showstoppers or intricacies that would be relevant for other people if I take too long.

------------

A simple test for direct left-recursion is the above mentioned PEG grammar translated to PyParsing:

```python3
import pyparsing as pp

E = pp.Forward("E")
num = pp.Word(pp.nums)
E <<= E + '+' - num | num

E.parseString("1+2")
```

**Comment by maxfischer2781** [comment]
A naive implementation seems to work. However, it is *not* a drop-in replacement for either of the existing parsing strategies, since it comes with its own memoization:
- It does not quite match Packrat, since memoization happens for ``Forward`` clauses and must be unbounded.
- It seems to have very long/infinite parse time for some tests when aggressively clearing the memoization cache.

The first is expected, but the second is surprising. I have not yet found out what is special about the failing tests; could be that they just do not work for memoization, or that they use some non-PEG operators.

**Comment by ptmcg** [comment]
This is very interesting work. I had downloaded Guido's article on how LR was handled in the Python PEG parser, but have not had a real chance to look at how it would relate to pyparsing's logic.

I was thinking that this might be something to save until after 3.0 goes out, since it is potentially disruptive to existing code. You  sound like you have made a lot of progress, I will look at your work as soon as I can.

Do the existing pyparsing tests also still pass with your changes?

**Comment by maxfischer2781** [comment]
> This is very interesting work. I had downloaded Guido's article on how LR was handled in the Python PEG parser, but have not had a real chance to look at how it would relate to pyparsing's logic.

I admit the paper was a bit convoluted for me and Guido's article seems to skip a problem that I ran into with another parser approach. Trying to put it into pyparsing was a practical approach to understanding it.  
So any comments on making the description/code easier to follow are highly welcome.

> Do the existing pyparsing tests also still pass with your changes?

In a way yes, since this needs to be switched on just like Packrat. But with recursion switched on they don't, since it's not a 100% percent drop-in replacement – it has to evaluate ``Forward`` clauses at least twice to check if they grow via recursion. So anything that relies on strict evaluation count/sequence via side-effects won't work the same.

**Comment by maxfischer2781** [comment]
It turns out the ``Or`` clause is tricky: It evaluates its subclauses in a loop without actions, which causes a rapid expansion in a single match. But once it has decided on a match, it re-matches it with actions on the initial expansion which is not expanded enough.

**Comment by maxfischer2781** [comment]
The other major question is how to handle empty matches.

For reference, the algorithm from literature insists on monotonic progress (E[n] > E[n-1]) to proceed. Given the ``Empty()``/``ε`` clause and the rule
```
E <- E '+' ε / num
```
Monotonic progress avoids that one can infinitely "increase" the match to ``num``, ``num ε``, ``num ε ε``, ...

Since PyParsing is not just a PEG parser but also supports actions, the situation is not equivalent: There are viable ``ε`` matches that get meaning through actions. For example, we could have the following rule with ``{...}`` actions
```
E <- E '+' ε { 2 } / E '+' ε { 1 } / num
```
Here, the actions capture the path through the grammar. While the match span (``loc``) does not increase, the match result ``num``, ``num 1``, ``num 2``, ``num 1 2`` differ significantly. And one could of course argue that the "correct" result is something like ``num 1 2 2 2 2 ...``.

---------------------------

In #288 I've settled for allowing a single empty match in the beginning. This supports non-recursive rules, like ``E <- ε { 1 }``, without turning to infinite recursion. It also comes naturally when the base case is a "negative" match (``loc - 1``) which is shorter than an empty match.

This seems to work fine so far, but perhaps there are cases for which this is too naive.

**Comment by maxfischer2781** [comment]
The issue with ``Or`` and issues with strict evaluation, i.e. each clause may be evaluated more than once, are related: they come down to how recursion interacts with actions. That is something a pure PEG *parser* does not have to concern itself with, but relevant for PyParsing.

-------------------

PyParsing can match with ``doActions=True`` and ``doActions=False``. The latter is useful to "peek" at a match without side-effects. For example, ``Or`` uses this to get the longest match of *all* matches before evaluating any.  
A naive implementation (like mine 😅) of the Medeiros algorithm merely matches a ``Forward`` element with whatever ``doActions`` it receives. This directly leads to two issues:
- The ``doActions=True`` and ``doActions=False`` paths can go out of sync, with one being matched fully while the other is still being matched. This is what breaks ``Or`` clauses.
- The ``doActions=True`` path will "over-evaluate", as it runs actions while peeking at the next match. This is what breaks any stack-based interpreter actions, like ``fourFn`` and indentation.

Simplified, the current algorithm looks like this:
```python3
old_match = NO_MATCH
while True:
    # try next recursion level with current settings
    new_match = element.parseImpl(instring, loc, doActions)
    if new_match > old_match:
        old_match = new_match
    else:
        return old_match
```

A solution is to evaluate both ``doActions=True`` and ``doActions=False`` together:
```python3
old_peek, match = NO_PEEK, NO_MATCH
while True:
    # peek at next recursion level without actions
    new_peek = element.parseImpl(instring, loc, False)
    if new_peek > old_peek:
        # match next recursion level with actions
        match = element.parseImpl(instring, loc, True)
    else:
        return match
```
This ensures the two paths do not go out of sync (if they get evaluated implicitly, they won't expand) and that there is no over-evaluation (an action is only run if the match is better).

-------------------

The simplification omits some simple chores (error handling) but the tricky part is 1) handling the memo and 2) handling the ``doActions=False`` part in isolation. I haven't tried it yet, but my assumptions are:
- There needs to be separate logic for the pure ``doActions=False`` path, which does not intermingle with the ``doActions=True`` path. This is to allow proper behaviour when we are *not* recursing, but another outer element peeks ahead.
- In the ``doActions=True`` path if there is no memo entry for ``doActions=True`` then discard any previous ``doActions=False`` paths and start again. This is to ensure both expand together.
- During expansion, only update the memo of both paths when the current level was matched for both paths. Otherwise, the paths slightly diverge if a called-on element peeks with ``doActions=False``.

**Comment by ptmcg** [comment]
Merged and pushed out in 3.0.0b3. Let's handle further LR business in a new issue.
